### ISSUE TITLE
Double filename extension not supported with EXTENSION_MAPPING

### ISSUE THREAD

**Issue by AlexisWilke** [doxygen/doxygen#7049]
Very often, people use an extra extension, `.in`, when they intend to have a `configure()` run against a file.

In CMake and `./configure` scripts, it is quite conventional to write a header file with many parameters defined as `@NAME@`. Such names get converted to values by CMake or the `./configure` script.

For example, in my case I convert a few `@...@` with my tool's  `version.hpp.in` file:

    #define    LPP_VERSION_MAJOR   @LPP_VERSION_MAJOR@
    #define    LPP_VERSION_MINOR   @LPP_VERSION_MINOR@
    #define    LPP_VERSION_PATCH   @LPP_VERSION_PATCH@
    #define    LPP_VERSION_BUILD   @LPP_VERSION_BUILD@
    #if LPP_VERSION_PATCH
    #define    LPP_VERSION_STRING  "@LPP_VERSION_MAJOR@.@LPP_VERSION_MINOR@.@LPP_VERSION_PATCH@"
    #else
    #define    LPP_VERSION_STRING  "@LPP_VERSION_MAJOR@.@LPP_VERSION_MINOR@"
    #endif

    int             get_major_version();
    int             get_release_version();
    int             get_patch_version();
    char const *    get_version_string();

And in the corresponding CMake file I have:

    configure_file(
        ${CMAKE_CURRENT_SOURCE_DIR}/version.hpp.in
        ${CMAKE_CURRENT_BINARY_DIR}/version.hpp
    )

This works like a charm and once compiled I get the version numbers in my header file, just as expected. We can see the result in the binary tree.

To get the documentation, I run Doxygen. Since the `.hpp.in` is not conventional, I try to add that extension to my doxy file like  so:

    EXTENSION_MAPPING      = h.in=C++ \
                             hpp.in=C++

    FILE_PATTERNS          = *.h \
                             *.h.in \
                             *.hpp \
                             *.hpp.in \
                             *.cpp \
                             *.c

The `FILE_PATTERNS` works since I see my `version.hpp.in` file (and in project where I use `version.h.in`, I see that file. So the `.h.in` and `.hpp`.in extensions work in the `FILE_PATTERNS`.)

However, nothing from that file gets documented. I have a file entry like so:

    /** \file
     * \brief Lalala
     *
     * Blah blah blah
     */

And the output totally ignores that. Doxygen ignores the `h.in=C++` specification.

As mentioned by someone on [Stackoverflow](https://stackoverflow.com/questions/56537009), it works if I have a single extension. So if I remove the second period I can make it work indeed. `hin=C++`. However, that's not the conventional extension.

Would it be possible for you to transform the code to support the equivalent of an "end with" instead of just taking the last extension of a filename?

What I would do is something like this:

    extension_mapping_ext = "." + extension_mapping_ext;
    for(auto f : files)
    {
        if(f->filename.length() > extension_mapping_ext.length()
        && f->filename.substr(f->filename.length() - extension_mapping_ext.length()) == extension_mapping_ext)
        {
            ...found a match!...
        }
    }

So in other words, I would do a form of `f->filename.ends_with(extension_mapping_ext)` after prepending a period to the `EXTENSION_MAPPING` extension and making sure that the filename is at least one more character than just the extension.

I was able to reproduce the problem with version 1.8.13 so I would imagine that the _bug_ is still present in the latest version.

**Comment by albert-github** [comment]
Some thoughts:

With a combination of `EXTENSION_MAPPING`, `INPUT` and `FILTER_PATTERNS` there are possibilities to use non default file extensions and translate parts to be useful in the documentation (here this would be for e.g. `@LPP_VERSION_MAJOR@`.
The problem occurs when the end of the filename is in the form of  `.xx.yy` i.e. consists of multiple parts.

The relevant question that pops up here is: **What is the extension in case of e.g. `zz.xx.yy`?**
 - the part of the filename from the last '.' till the end
 - the part of the filename from the first '.' till the end

another possibility might be to have an extended definition where giving a proposed extension (which starts with a '.', in case of e.g. `EXTENSION_MAPPING` this can be automatically added as it is done now too)  it is checked against the last part of the filename if it matches, so when having the proposed extension `.bb.aa` it would match:
- `cc.bb.aa` and giving as basename `cc`
- `dd.cc.bb.aa` and giving as basename `dd.cc`

At the moment I see on a number of placed a `findRev('.')` and also `find('.'), most of them related to filenames, but some are also related to e.g. class members. It might even be the fact that not all extensions and basenames are "calculated" correctly.


Note:
Looking at the Qt documentation (https://doc.qt.io/qt-5/qfileinfo.html) there are the methods (these are NOT used by doxygen! but given as a bit of a reference):

> QString QFileInfo::suffix() const
>   Returns the suffix (extension) of the file.
>   The suffix consists of all characters in the file after (but not including) the last '.'.

> QString QFileInfo::completeSuffix() const
>   Returns the complete suffix (extension) of the file.
>   The complete suffix consists of all characters in the file after (but not including) the first '.'.

> QString QFileInfo::baseName() const
> Returns the base name of the file without the path.
> The base name consists of all characters in the file up to (but not including) the first '.' character.

> QString QFileInfo::completeBaseName() const
> Returns the complete base name of the file without the path.
> The complete base name consists of all characters in the file up to (but not including) the last '.' character.

The major question here is that for this enhancement request which definition should be used by doxygen (@doxygen?)

**Comment by asmaloney** [comment]
Just a (necro-)ping that I ran into this exact problem (.h.in used in configure() in CMake).

I set up the `EXTENSION_MAPPING` mapping `h.in=C++` and couldn't figure out why it wasn't working!

> The major question here is that for this enhancement request which definition should be used by doxygen?

I would have expected it to match everything that ends with `.h.in`.
